app/components/code-mirror.js

```diff
@@ -0,0 +1,188 @@
+  highlightTrailingWhitespace: (enabled) => (enabled ? [highlightTrailingWhitespace()] : []),
+  highlightWhitespace: (enabled) => (enabled ? [highlightWhitespace()] : []),
+  history: (enabled) => (enabled ? [history(), keymap.of(historyKeymap)] : []),
+  indentOnInput: (enabled) => (enabled ? [indentOnInput()] : []),
+  indentUnit: (indentUnitText) => (indentUnitText !== undefined ? [indentUnit.of(indentUnitText)] : []),
+  indentWithTab: (enabled) => (enabled ? [keymap.of([indentWithTab])] : []),
+  lineNumbers: (enabled) => (enabled ? [lineNumbers()] : []),
+  lineSeparator: (lineSeparatorText) => (lineSeparatorText !== undefined ? [EditorState.lineSeparator.of(lineSeparatorText)] : []),
+  lineWrapping: (enabled) => (enabled ? [EditorView.lineWrapping] : []),
+  placeholder: (placeholderText) => (placeholderText !== undefined ? [placeholder(placeholderText)] : []),
+  readOnly: (enabled) => [EditorState.readOnly.of(enabled)],
+  rectangularSelection: (enabled) => (enabled ? [rectangularSelection()] : []),
+  scrollPastEnd: (enabled) => (enabled ? [scrollPastEnd()] : []),
+  syntaxHighlighting: (enabled) => (enabled ? [syntaxHighlighting(defaultHighlightStyle, { fallback: true })] : []),
+  tabSize: (enabled) => (enabled !== undefined ? [EditorState.tabSize.of(enabled)] : []),
+  theme: (theme) => (theme !== undefined && SUPPORTED_THEMES.includes(theme) ? [THEME_EXTENSIONS[theme]] : []),
+  // advanced handling
+  languageOrFilename: async (newValue, { language, filename }) => {
+    const detectedLanguage = language
+      ? LanguageDescription.matchLanguageName(languages, language)
+      : filename
+        ? LanguageDescription.matchFilename(languages, filename)
+        : undefined;
+
+    let loadedLanguage = [];
+
+    if (detectedLanguage) {
+      switch (detectedLanguage.name.toLowerCase()) {
+        case 'markdown':
+          loadedLanguage = markdown({ codeLanguages: languages });
+          break;
+        default:
+          loadedLanguage = await detectedLanguage.load();
+          break;
+      }
+    }
+
+    return loadedLanguage ? [loadedLanguage] : [];
+  },
+  originalDocumentOrMergeControls: (newValue, { originalDocument, mergeControls }) => {
+    return originalDocument
+      ? [
+          unifiedMergeView({
+            original: originalDocument,
+            mergeControls: !!mergeControls,
+            // collapseUnchanged: { margin: 3, minSize: 4 },
+          }),
+        ]
+      : [];
+  },
+};
+
+export default class CodeMirrorComponent extends Component {
+  renderedView;
+
+  compartments = Object.keys(OPTION_HANDLERS).reduce(function (accumulator, optionName) {
+    accumulator[optionName] = new Compartment();
+
+    return accumulator;
+  }, {});
+
+  @action documentDidChange(element, [newValue]) {
+    const documentChanged = this.renderedView.state.doc.toString() !== newValue;
+
+    if (!documentChanged) {
+      return;
+    }
+
+    if (this.args.history && !this.args.preserveHistory) {
```

Comment:
Note: I'm not looking at the naming at the moment, will take a look again once the full interface is present as a TS type. From a quick look, this doesn't match our usual naming standards - for example, here the name `history` is pretty ambiguous. Is this a boolean? Is it some kind of object? If this is a boolean, we'd typically name it something like `shouldRecordHistory` (or something like that).

Reply:
If the intention here is to map 1:1 to codemirror names, then _maybe_ that's okay - maybe the added indirection of mapping these to our naming standard isn't worth it 🤔

Reply:
The naming currently matches the names of CodeMirror extensions